src/expr/src/relation/mod.rs

```diff
@@ -1784,6 +1784,39 @@ impl MirRelationExpr {
         dfs(self, &mut size, &mut max_depth, 1);
         (size, max_depth)
     }
+
+    /// Returns whether a expression can run an expensive function. We err on returning false.
+    /// The goal is to be a heuristic for "this expression is cheap to evaluate:
+    /// ideally not creating a dataflow at all when optimized".
+    pub fn can_run_expensive_function(&self) -> bool {
+        /// Returns whether a scalar expression can run an expensive function
+        fn scalar_can_run_expensive_function(scalar: &MirScalarExpr) -> bool {
+            match scalar {
+                MirScalarExpr::Column(_)
+                | MirScalarExpr::Literal(_, _)
+                | MirScalarExpr::CallUnmaterializable(_) => false,
+                MirScalarExpr::CallUnary { .. }
+                | MirScalarExpr::CallBinary { .. }
+                | MirScalarExpr::CallVariadic { .. } => true,
+                MirScalarExpr::If { cond, then, els } => {
+                    scalar_can_run_expensive_function(cond)
+                        || scalar_can_run_expensive_function(then)
+                        || scalar_can_run_expensive_function(els)
+                }
+            }
+        }
+
+        // FlapMap has a table function while all other constructs use MirScalarExpr to run a function
+        let mut result = match self {
+            MirRelationExpr::FlatMap { .. } => true,
+            _ => false,
+        };
+        self.visit_scalars(&mut |scalar| {
+            result = result || scalar_can_run_expensive_function(scalar)
+        });
+        self.visit_children(|e| result = result || e.can_run_expensive_function());
```

Comment:
This looks like it ignores expensive joins. `select * from a, b`. Am I reading this correctly? Is this ok?

Reply:
That's correct. I think it's fine for now: prohibiting joins would change the space of "simple" queries but I don't think we can ever get to a definition (besides an allowlist) that doesn't let in expensive queries. Since this is only checked when there are no dataflow dependencies the underlying relations being joined shouldn't be too large

Reply:
Ah, I didn't connect the other logic above that excludes things with dependencies.